File: app/src/main/java/com/johnpetitto/orachat/ui/chats/ChatsAdapter.java
package com.johnpetitto.orachat.ui.chats;

import android.content.Context;
import android.support.v7.widget.RecyclerView;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.TextView;

import com.johnpetitto.orachat.R;
import com.johnpetitto.orachat.TimeUtilsKt;
import com.johnpetitto.orachat.data.chat.Chat;
import com.johnpetitto.orachat.data.chat.ChatMessage;
import com.johnpetitto.orachat.data.user.User;

import java.util.ArrayList;
import java.util.List;

import butterknife.BindView;
import butterknife.ButterKnife;

public class ChatsAdapter extends RecyclerView.Adapter<RecyclerView.ViewHolder> {
    private static final int HEADER_TYPE = 0;
    private static final int ITEM_TYPE = 1;

    private List<Object> data = new ArrayList<>();
    private OnChatClickListener listener;

    public interface OnChatClickListener {
        void onChatClick(Chat chat);
    }

    public ChatsAdapter(OnChatClickListener listener) {
        this.listener = listener;
    }

    @Override
    public RecyclerView.ViewHolder onCreateViewHolder(ViewGroup parent, int viewType) {
        RecyclerView.ViewHolder holder = null;
        LayoutInflater inflater = LayoutInflater.from(parent.getContext());

        switch (viewType) {
            case HEADER_TYPE:
                holder = new VHHeader(inflater.inflate(R.layout.list_header_chat, parent, false));
                break;
            case ITEM_TYPE:
                holder = new VHItem(inflater.inflate(R.layout.list_item_chat, parent, false));
                break;
        }

        return holder;
    }

    @Override
    public void onBindViewHolder(RecyclerView.ViewHolder holder, int position) {
        switch (holder.getItemViewType()) {
            case HEADER_TYPE:
                String date = (String) data.get(position);
                TextView header = (TextView) holder.itemView;

                if (date.equals("Today")) {
                    header.setText(R.string.today);
                } else {
                    header.setText(date);
                }

                break;
            case ITEM_TYPE:
                Chat chat = (Chat) data.get(position);
                VHItem item = (VHItem) holder;
                Context context = item.itemView.getContext();

                User chatOwner = chat.getUsers().get(0); // TODO: is first user the owner?
                String formattedName = context.getString(R.string.chat_item_name, chat.getName(), chatOwner.getName());
                item.name.setText(formattedName);

                ChatMessage lastMessage = chat.getLastChatMessage();
                String lastUserName = lastMessage.getUser().getName();
                String timeAgo = TimeUtilsKt.getTimeAgo(context, lastMessage.getCreatedAt());
                String formattedLastUpdate = context.getString(R.string.chat_message_timestamp, lastUserName, timeAgo);
                item.lastUpdate.setText(formattedLastUpdate);

                item.lastMessage.setText(lastMessage.getMessage());

                break;
        }
    }

    @Override
    public int getItemCount() {
        return data.size();
    }

    @Override
    public int getItemViewType(int position) {
        return data.get(position) instanceof String ? HEADER_TYPE : ITEM_TYPE;
    }

    public void setChatsGroupedByDate(List<Object> chatsGroupedByDate) {
        data = chatsGroupedByDate;
        notifyDataSetChanged();
    }

    public void addChatsGroupedByDate(List<Object> chatsGroupedByDate) {
        int start = data.size();
        data.addAll(chatsGroupedByDate);
        notifyItemRangeInserted(start, chatsGroupedByDate.size());
    }

    @SuppressWarnings("WeakerAccess")
    class VHHeader extends RecyclerView.ViewHolder {
        VHHeader(View itemView) {
            super(itemView);
        }
    }

    class VHItem extends RecyclerView.ViewHolder {
        @BindView(R.id.chats_item_name) TextView name;
        @BindView(R.id.chats_item_last_update) TextView lastUpdate;
        @BindView(R.id.chats_item_last_message) TextView lastMessage;

        VHItem(View itemView) {
            super(itemView);
            ButterKnife.bind(this, itemView);
            itemView.setOnClickListener(new View.OnClickListener() {
                @Override
                public void onClick(View view) {
                    Chat chat = (Chat) data.get(getAdapterPosition());
                    listener.onChatClick(chat);
                }
            });
        }
    }
}

File: app/src/main/java/com/johnpetitto/orachat/ui/chats/ChatsAdapter.kt
package com.johnpetitto.orachat.ui.chats

import android.support.v7.widget.RecyclerView
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import com.johnpetitto.orachat.R
import com.johnpetitto.orachat.data.chat.Chat
import com.johnpetitto.orachat.getTimeAgo
import kotlinx.android.synthetic.main.list_header_chat.view.*
import kotlinx.android.synthetic.main.list_item_chat.view.*

private const val HEADER_TYPE = 0
private const val ITEM_TYPE = 1

class ChatsAdapter(
        private val listener: OnChatClickListener,
        private var chatData: MutableList<Any> = mutableListOf()):
        RecyclerView.Adapter<RecyclerView.ViewHolder>()
{
    interface OnChatClickListener {
        fun onChatClick(chat: Chat)
    }

    override fun getItemCount() = chatData.size

    override fun onCreateViewHolder(parent: ViewGroup?, viewType: Int): RecyclerView.ViewHolder {
        val inflater = LayoutInflater.from(parent!!.context)
        return when (viewType) {
            HEADER_TYPE -> VHHeader(inflater.inflate(R.layout.list_header_chat, parent, false))
            else -> VHItem(inflater.inflate(R.layout.list_item_chat, parent, false))
        }
    }

    override fun onBindViewHolder(holder: RecyclerView.ViewHolder?, position: Int) {
        when (holder!!.itemViewType) {
            HEADER_TYPE -> (holder as VHHeader).bind(chatData[position] as String)
            ITEM_TYPE -> (holder as VHItem).bind(chatData[position] as Chat)
        }
    }

    override fun getItemViewType(position: Int) = when(chatData[position]) {
        is String -> HEADER_TYPE
        else -> ITEM_TYPE
    }

    fun setChatData(chatData: List<Any>) {
        this.chatData.clear()
        this.chatData.addAll(chatData)
        notifyDataSetChanged()
    }

    fun addChatData(chatData: List<Any>) {
        val start = this.chatData.size
        this.chatData.addAll(chatData)
        notifyItemRangeChanged(start, chatData.size)
    }

    inner class VHHeader(itemView: View): RecyclerView.ViewHolder(itemView) {
        fun bind(header: String) {
            itemView.headerLabel.text = if (header == "Today") {
                itemView.context.getString(R.string.today)
            } else {
                header
            }
        }
    }

    inner class VHItem(itemView: View): RecyclerView.ViewHolder(itemView) {
        fun bind(chat: Chat) {
            val ctx = itemView.context

            val ownerName = chat.users.first().name
            val chatName = ctx.getString(R.string.chat_item_name, chat.name, ownerName)
            itemView.chatName.text = chatName

            val lastMessage = chat.lastChatMessage
            val lastUserName = lastMessage.user.name
            val timeAgo = getTimeAgo(ctx, lastMessage.createdAt)
            val lastUpdate = ctx.getString(R.string.chat_message_timestamp, lastUserName, timeAgo)
            itemView.lastUpdate.text = lastUpdate
            itemView.lastMessage.text = lastMessage.message

            itemView.setOnClickListener { listener.onChatClick(chatData[adapterPosition] as Chat) }
        }
    }
}
